File: fabric-example-mod-26.2/src/client/java/dev/pvpbot/net/Activations.java
package dev.pvpbot.net;

/** Активации, реализованные чистой Java (без библиотек). */
public final class Activations {
    public static float relu(float x) { return Math.max(0f, x); }
    public static float tanh(float x) { return (float) Math.tanh(x); }
    public static float sigmoid(float x) {
        if (x >= 0f) return (float) (1.0 / (1.0 + Math.exp(-x)));
        float e = (float) Math.exp(x);
        return e / (1.0f + e);
    }
    public static float linear(float x) { return x; }

    public static float apply(String name, float x) {
        switch (name) {
            case "relu":   return relu(x);
            case "tanh":   return tanh(x);
            case "sigmoid":return sigmoid(x);
            default:       return linear(x);
        }
    }

    private Activations() {}
}

File: training/pvp_bot/activations.py
"""Активации и их производные. Чистый numpy, никаких ML-фреймворков."""
import numpy as np


def relu(x):
    return np.maximum(0.0, x)


def relu_grad(x):
    return (x > 0).astype(np.float64)


def tanh(x):
    return np.tanh(x)


def tanh_grad(x):
    return 1.0 - np.tanh(x) ** 2


def sigmoid(x):
    # стабильная версия, чтобы не было переполнения
    out = np.empty_like(x, dtype=np.float64)
    pos = x >= 0
    out[pos] = 1.0 / (1.0 + np.exp(-x[pos]))
    neg = ~pos
    e = np.exp(x[neg])
    out[neg] = e / (1.0 + e)
    return out


def sigmoid_grad(x):
    s = sigmoid(x)
    return s * (1.0 - s)


def linear(x):
    return x


def linear_grad(x):
    return np.ones_like(x, dtype=np.float64)


_ACT = {
    "relu": (relu, relu_grad),
    "tanh": (tanh, tanh_grad),
    "sigmoid": (sigmoid, sigmoid_grad),
    "linear": (linear, linear_grad),
}


def forward(name, x):
    return _ACT[name][0](x)


def grad(name, x):
    return _ACT[name][1](x)
